packages/tableofcontents.lua

```diff
@@ -159,8 +92,96 @@ SILE.doTexlike([[%
 \define[command=tableofcontents:level2number]{}%
 \define[command=tableofcontents:level3number]{}%
 ]])
+end)
+
+end
 
-  end,
+local function registerCommands (class)
+
+  SILE.registerCommand("tableofcontents", function (options, _)
+    local depth = SU.cast("integer", options.depth or 3)
+    local linking = SU.boolean(options.linking, true)
+    local tocfile,_ = io.open(SILE.masterFilename .. '.toc')
+    if not tocfile then
+      SILE.call("tableofcontents:notocmessage")
+      return
+    end
+    local doc = tocfile:read("*all")
+    local toc = assert(load(doc))()
+    SILE.call("tableofcontents:header")
+    for i = 1, #toc do
+      local item = toc[i]
+      if item.level <= depth then
+        SILE.call("tableofcontents:item", {
+          level = item.level,
+          pageno = item.pageno,
+          number = item.number,
+          link = linking and item.link
+        }, item.label)
+      end
+    end
+    SILE.call("tableofcontents:footer")
+    _tableofcontents = toc
```

Comment:
Somewhat unrelated, but since you are on refactoring things here and adding hooks, etc...

Since `_tableofcontents` is a local variable, it means that one cannot easily just redefine the tableofcontents command to process and display it (possibly very) differently. I had, typically this problem in my "fancytoc" package where I just want to show the TOC in a different way (= hierarchical, with braces, rather than line by line).
There could be two approaches:
- Moving that to a scratch global variable (though it does expose some of the internal trickery) so that a package wishing to overwrite the `\tableofcontent` command could fully reimplement it.
- Or split the whole TOC loading in an exported function. I used that strategy for my fancytoc package in my repo, not yet in main but [here it is for the record](https://github.com/Omikhleia/omikhleia-sile-packages/blob/book-project-iteration/packages/experimental/fancytoc.lua)

Reply:
I agree this is less than ideal but am going to leave it for a future refactor to take care of. We have a huge mix of private module vars vs. random scratch globals with inconsistent names. Also it isn't clear what things in scratch are "public" in the sense that changing our use of them should be considered breaking vs. those being private implementation details to use at your own risk.

Reply:
c.f. #1417